python/cudf_polars/cudf_polars/dsl/ir.py

```diff
@@ -802,6 +812,61 @@ def evaluate(self, *, cache: MutableMapping[int, DataFrame]) -> DataFrame:
         return DataFrame(broadcasted).slice(self.options.slice)
 
 
+class ConditionalJoin(IR):
+    """A conditional inner join of two dataframes on a predicate."""
+
+    __slots__ = ("predicate", "options", "ast_predicate")
+    _non_child = ("schema", "predicate", "options")
+    predicate: expr.Expr
+    options: tuple
+
+    def __init__(
+        self, schema: Schema, predicate: expr.Expr, options: tuple, left: IR, right: IR
+    ) -> None:
+        self.schema = schema
+        self.predicate = predicate
+        self.options = options
+        self.children = (left, right)
+        self.ast_predicate = to_ast(predicate)
+        _, join_nulls, _, _, coalesce = self.options
+        # Preconditions from polars
+        assert not join_nulls
+        assert not coalesce
+        if self.ast_predicate is None:
+            raise NotImplementedError(
+                f"Conditional join with predicate {predicate}"
+            )  # pragma: no cover; polars never delivers expressions we can't handle
+
+    def evaluate(self, *, cache: MutableMapping[int, DataFrame]) -> DataFrame:
+        """Evaluate and return a dataframe."""
+        left, right = (c.evaluate(cache=cache) for c in self.children)
+        _, _, zlice, suffix, _ = self.options
+        lg, rg = plc.join.conditional_inner_join(
+            left.table, right.table, self.ast_predicate
+        )
```

Comment:
Polars only supports inner joins for its inequality joins right now, so I didn't refactor to share code.